### PR TITLE
executor,planner: fix update join update unmatched outer row

### LINKED ISSUE #23482: update right columns in left join report wrong null check error when no match record in right side
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
func (s *testSuite6) TestUpdateClusteredJoin(c *C) {
	tk := testkit.NewTestKit(c, s.store)
	tk.MustExec("use test")
	tk.MustExec("drop table if exists a, b")
	tk.MustExec("create table a (k1 int, k2 int, v int)")
	tk.MustExec("create table b (a int, k1 int, k2 int, v int, primary key(k1, k2))")
	tk.MustExec("insert into a values (1, 1, 1)")
	tk.MustExec("update a left join b on a.k1 = b.k1 and a.k2 = b.k2 set a.v = 2, b.v = 100")
	tk.MustQuery("select * from b").Check(testkit.Rows())
	tk.MustQuery("select * from a").Check(testkit.Rows("1 1 2"))
}
```

### 2. What did you expect to see? (Required)

success

### 3. What did you see instead (Required)

```
[table:1048]Column 'k1' cannot be null ("[table:1048]Column 'k1' cannot be null")
```

without not-null  constraint, it will write down wrong null row data in table `b`

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve?

Issue Number: close #23482 

Problem Summary:

- 658132fcf3d859204ec53c0caaf27d837a63cf00 make non-clustered table's outer row be update
- clustered index implement has some TODO indicate that not be right handle

### What is changed and how it works?

What's Changed, How it Works:

- in `UpdateExec.prepare`(called for each row) set `UpdateExec.updatable` with right value
- choose any one column in CommonHandle to check `is it NULL row generated by unmatched outer row`  

### Related changes

- Need to cherry-pick to the release branch 5.0

### Check List 

Tests 

- Unit test

Side effects

- n/a

### Release note 

- No release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/23491)

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -4172,8 +4171,8 @@ func (c TblColPosInfoSlice) Less(i, j int) bool {
 	return c[i].Start < c[j].Start
 }
 
-// FindHandle finds the ordinal of the corresponding handle column.
-func (c TblColPosInfoSlice) FindHandle(colOrdinal int) (int, bool) {
+// FindTblColPosIdx finds the ordinal of the corresponding access column.
+func (c TblColPosInfoSlice) FindTblColPosIdx(colOrdinal int) (int, bool) {
```
Comment (coocood) [inline planner/core/logical_plan_builder.go:4175]: I think it's more accurate to name it `FindTableIdx`.

File: executor/update.go [hunk 2]
```diff
@@ -96,6 +97,9 @@ func (e *UpdateExec) prepare(ctx context.Context, schema *expression.Schema, row
 				break
 			}
 		}
+		if e.unmatchedOuterRow(content, row) {
+			updatable = false
+		}
 		e.updatable = append(e.updatable, updatable)
```
Comment (coocood) [inline executor/update.go:103]: rename it to `tableUpdatable` is easier to understand.

**Comment by lysu** [comment]
need add more swap value test case

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -4337,46 +4331,28 @@ func (b *PlanBuilder) buildUpdate(ctx context.Context, update *ast.UpdateStmt) (
...
 	pkUpdated bool
 }
 
-func checkUpdateList(ctx sessionctx.Context, tblID2table map[int64]table.Table, updt *Update) error {
-	assignFlags, err := GetUpdateColumns(ctx, updt.OrderedList, updt.SelectPlan.Schema().Len())
-	if err != nil {
-		return err
-	}
+// CheckUpdateList checks all related columns in updatable state.
+func CheckUpdateList(assignFlags []int, updt *Update) error {
```
Comment (jackysp) [inline planner/core/logical_plan_builder.go:4345]: how about move it from `plancore` to `executor`?

File: planner/core/logical_plan_builder.go [hunk 4]
```diff
@@ -4337,46 +4331,28 @@ func (b *PlanBuilder) buildUpdate(ctx context.Context, update *ast.UpdateStmt) (
 		tblID2table[id], _ = b.is.TableByID(id)
 	}
 	updt.TblColPosInfos, err = buildColumns2Handle(updt.OutputNames(), tblID2Handle, tblID2table, true)
-	if err == nil {
-		err = checkUpdateList(b.ctx, tblID2table, updt)
-	}
 	updt.PartitionedTable = b.partitionedTable
+	updt.tblID2Table = tblID2table
```
Comment (dyzsr) [inline planner/core/logical_plan_builder.go:4335]: `tblID2Table` is built twice, once in `PlanBuilder.buildUpdate` building from `tblID2Handle`, another is in `executorBuilder.buildUpdate` building from `TblColPosInfos`. And they are quite similar.

Can we remove the one in `Update` plan, if the only usage of `Update.tblID2Table` is for `CheckUpdateList`?

Comment (lysu) [reply planner/core/logical_plan_builder.go:4335]: @dyzsr great idea, but it seems executorBuilder.buildUpdate take care more about partitions?

how about add another issue to improve it? this's PR is release-block bugfix, introduce addition modification is high risk

**Comment by lysu** [comment]
@coocood @dyzsr  PTAL

File: executor/builder.go [hunk 5]
```diff
@@ -1842,6 +1842,15 @@ func (b *executorBuilder) buildUpdate(v *plannercore.Update) Executor {
 	}
 	base := newBaseExecutor(b.ctx, v.Schema(), v.ID(), selExec)
 	base.initCap = chunk.ZeroCapacity
+	var assignFlag []int
+	assignFlag, b.err = getAssignFlag(b.ctx, v, base.children[0].Schema().Len())
```
Comment (tangenta) [inline executor/builder.go:1846]:
```suggestion
	assignFlag, b.err = getAssignFlag(b.ctx, v, selExec.Schema().Len())
```

**Review by tangenta — APPROVED** [review]

**Pull request merged**